### ISSUE TITLE
F401 triggers based on renaming imports

### ISSUE THREAD

**Issue by henryiii** [astral-sh/ruff#7303]
Ruff's F401 triggers if you import something _and_ rename it, but not if you keep the name the same.

For example, this file:

```python
from . import clients as clients
from .config import config as config
from .config_training import config as config_training # E: F401 [*] `.config_training.config` imported but unused
from .thing import other as foobar # E: F401 [*] `.thing.other` imported but unused
from . import x as y # E: F401 [*] `.x` imported but unused
from a import b as c # E: F401 [*] `a.b` imported but unused
```

Which I'm running in VI via ALE using the LSP to get the comments, but I can also run it manually:

```console
$ ruff check --isolated --select=F401 example.py
example.py:3:40: F401 [*] `.config_training.config` imported but unused
example.py:4:29: F401 [*] `.thing.other` imported but unused
example.py:5:20: F401 [*] `.x` imported but unused
example.py:6:20: F401 [*] `a.b` imported but unused
Found 4 errors.
[*] 4 potentially fixable with the --fix option.
```

Notice that the error doesn't trigger if you `from an import b as b`, but only if you use `from an import b as c`. This was from an `__init__.py` file that didn't set `__all__`, which [isn't a good pattern](https://learn.scientific-python.org/development/patterns/exports/), and it goes away as it should if you set `__all__`). But I think it should be consistent. I'm pretty sure the problem is that `from a import b as b` is being treated as `from a import b; b = b` which "uses" `b`, making it not unused. That's a guess.

**Comment by zanieb** [comment]
We allow `from a import b as b` because it's respected by type checkers as a way to "export" the name similar to `__all__`. See the [Pyright documentation section](https://github.com/microsoft/pyright/blob/ada1ea63e8360fe3bd203bff3b0069c34d142817/docs/typed-libraries.md#library-interface) that discusses redundant aliases.

**Comment by Avasam** [comment]
This looks correct and intended. Whilst both import examples lead to unused symbol in the **current** module, both also lead to symbols that **exist**.
`from example import clients` and `from example import y` are both valid at runtime.
To keep public interfaces sane, type checkers (and some import linters like pycln, isort, etc.) have a bit of heuristic to define what is private and public, and the form `import X as X` is how you explicitly flag an import as an intended re-export.

**Comment by Avasam** [comment]
That being said, mixing explicit re-exports and `__all__` isn't a great pattern, and maybe a rule dedicated to that might satisfy your needs and expectations?

**Comment by henryiii** [comment]
Okay, I see why it happens, then, but the code that someone originally was confused about:

```python
from .config import config as config
from .config_training import config as config_training
```

still seems surprising - you can't "reexport" the second `config`, since it can't keep the original name. The current form obviously was trying to reexport, but if it has to be renamed, that doesn't work then.

I'll just suggest that they use `__all__`, explicit is better than implicit. And makes the imports look less odd. :)

**Comment by henryiii** [comment]
Thanks, by the way!

**Comment by henryiii** [comment]
> maybe a rule dedicated to that might satisfy your needs and expectations?

I was thinking the best solution would be to not do the explicit reexports, but move entirely to `__all__`.

I wouldn't mind having a rule that enforced all files (that don't start with an underscore, anyway) have an `__all__`, and all non-`__init__.py` files have a `def __dir__()` too (actually, I already enforce exactly this in pytest for scikit-build-core). But that's pretty specific to following this pattern.
